kep/399-sysched-scoring/README.md

```diff
@@ -0,0 +1,217 @@
+
+Initially, to schedule pod $P_1$, the ExS  scores for both Node 1 and Node 2 are the same (i.e., 0), since there are no pods running on the nodes (Figure 3(a)). Thus, we randomly pick one of the two nodes, Node 1 in this case, and updates Node 1's system call usage list using $P_1$'s system call profile (Figure 3(b)). When pod $P_2$ arrives to be scheduled, our plugin computes the ExS scores 7 and 0 for $P_2$ w.r.t. Node 1 and Node 2, respectively (Figure 3(b)). In this case, the plugins ranked Node 2 over Node 1 since Node 2 has a lower ExS  score (i.e., higher normalized score). Figure 3(c) shows the updated system call lists for Node 2. This process repeats for all incoming pods. After all pods are placed, the cluster-wide ExS score in this example is 12, i.e., $ExS_1$ is 5 and $ExS_2$ is 7, where 1 and 2 are node numbers.
+
+<p align="center">
+<img src="images/example_complete.png" style="width:50%">
+</p>
+
+<p align = "center">
+<b>Figure 3. Scheduling algorithm and pod placement example</b>
+</p>
+
+Figure 3 also highlights the opportunities not only for reducing the exposure to extraneous system calls (i.e., ExS score) but also the node’s attack surface. If $P_3$ had been scheduled on Node 2 instead of Node 1, then Node 2 would have a total of 7 system calls "open" $(1,3,4,5,7,8,9)$, instead of only 3 system calls under the SySched approach. 
+
+### Scheduling Plugin
+
+There are two main components of our plugin: 1) a lightweight in-memory cache that is kept synchronize with pod events in order to keep an up-to-date node to pod mapping in the cluster, and 2) mechanisms for retrieving the system call sets used by pods for computing the ExS scores. The in-memory cache reduces the cost of querying the API server for each pod scheduling event. We obtain the system call profiles for pods from their seccomp profiles. Our plugin leverages the Kubernetes sigs community-developed [Security Profiles Operator](https://github.com/kubernetes-sigs/security-profiles-operator) (SPO) that creates Custom Resource Definitions (CRDs) for seccomp profiles. If a pod has no seccomp profile associated with it, then our plugin simply returns, effectively a no-op.
+
+#### **How the Scheduler gets access to pods' system call profiles**
+
+We assume that the system call profiles of pods have explicitly been made available to the scheduler via the SPO. We rely on the SPO to generate and bind seccomp profile CRDs to the pods. The scheduler uses the API server to access a pod's seccomp profile CRD and parses the seccomp profile CRD to obtain the list of system calls for a pod. 
+
+An example of an SPO's seccomp profile CRD:
+
+```
+---
+apiVersion: security-profiles-operator.x-k8s.io/v1beta1
+kind: SeccompProfile
```

Comment:
We have multiple components of the seccomp profile to be taken into account depending on how the scoring calculation works in detail:

- A default action of `SCMP_ACT_ALLOW` may require us to collect / assume all available syscalls on a node
- The architecture has to match
- Additional arguments may not be allowed (it would be non-trivial to evaluate those)

Reply:
These are great points. 

Right now we are just considering `SCMP_ACT_ALLOW` but should definitely also handle other action types when they are present. Syscall arguments are not something we are considering using as of yet. Ultimately, we just care about what syscalls a pod is allowed to use. 

We will update these points in caveats/notes/limitations.